Fix issue 50611. Incorrect handling of K_FOREVER timeout in k_heap_al…

jshoen opened the pull request:
…igned_alloc.

Signed-off-by: Jay Shoen <jay.shoen@perceive.io>

Fixes #50611

fabiobaltieri commented:
Thanks for opening the PR @jshoen, can you redo the commit message format according to https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-guidelines?

Use amend to change it and then force push the update, see https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribution-workflow for more details.

jshoen commented:
Commit message amended.  Let me know if anything further is needed.

fabiobaltieri commented:
Hey, can you drop the merge commit? We don't use those, just reset --hard back to previous commit and force push again.

andyross commented on kernel/kheap.c:
```diff
@@ -77,9 +77,8 @@ void *k_heap_aligned_alloc(struct k_heap *h, size_t align, size_t bytes,
-		now = sys_clock_tick_get();
 		if (!IS_ENABLED(CONFIG_MULTITHREADING) ||
-		    (ret != NULL) || ((end - now) <= 0)) {
+		    (ret != NULL) || K_TIMEOUT_EQ(timeout, K_NO_WAIT)) {
```
This bit seems unrelated and AFAICT is a noop.  There is K_NO_WAIT handling in sys_clock_timeout_end_calc() already.

andyross replied:
Oh, actually this is a bug.  You can have "end < now" as a perfectly valid state, either because of a race between the two tick calls involved or because the upstream caller passed an absolute timeout in the past.  This code will then pend incorrectly in that case.

jshoen commented:
OK.  I just force pushed my original branch since I hadn't pulled down locally yet.  I think that should have done it.  I saw the message that the branch is out of date and needed merged, so I pushed the merge button.  I'll ignore in the future.

andyross commented on kernel/kheap.c:
```diff
@@ -93,9 +92,17 @@ void *k_heap_aligned_alloc(struct k_heap *h, size_t align, size_t bytes,
 		key = k_spin_lock(&h->lock);
+
+		if (!K_TIMEOUT_EQ(timeout, K_FOREVER)) {
+			int64_t remaining = end - sys_clock_tick_get();
```
Not loving the performance regression here: this is going to call back into the timer driver to recompute the next timeout at the end of the loop, even though we may very well succeed at the next allocation and not need the timeout.  It's also doing that math while holding the heap lock, which is needless (it's not touching heap data here).

Maybe what you want to do is retain the original structure and just fix the `end` value?  Something like this before the pend call?

```c
end = timeout == K_FOREVER ? INT64_MAX : end;
```

andyross replied:
Or a hybrid: retain your optimization of "skip the sys_clock_tick_get() if K_FOREVER", but do it before the pend instead of after.

jshoen replied:
OK.  I think that needs to be before the loop or the initial check will fail.

jshoen replied:
It is handy though on a low memory embedded platform.

jshoen replied:
Rather than
`end = timeout == K_FOREVER ? INT64_MAX : end;`
I could do:
`end &= INT64_MAX;`
Then there's no branch.  Do you have a preference?

andyross replied:
I guess a UB pedant would say that the latter is making assumptions about 2's complement representations, but I'm not one.  No preference here.

jshoen replied:
Went with your original proposal.  I think it's a little more clear what is being done.

andyross reviewed (changes requested):
Mostly just aesthetic/microperformance concerns, but this code is my baby so I'm allowed. :)

fabiobaltieri commented:
> OK. I just force pushed my original branch since I hadn't pulled down locally yet. I think that should have done it. I saw the message that the branch is out of date and needed merged, so I pushed the merge button. I'll ignore in the future.

Yup no worries, happens quite often, format is good now will leave to Andy and the others for the important stuff. :-)

andyross reviewed (dismissed):
Looks great, thanks.

mmahadevan108 commented:
@jshoen, please help fix these CI failure:
```
-:10: ERROR:CODE_INDENT: code indent should use tabs where possible
#10: FILE: kernel/kheap.c:70:
+        end = K_TIMEOUT_EQ(timeout, K_FOREVER) ? INT64_MAX : end;$

-:10: WARNING:LEADING_SPACE: please, no spaces at the start of a line
#10: FILE: kernel/kheap.c:70:
+        end = K_TIMEOUT_EQ(timeout, K_FOREVER) ? INT64_MAX : end;$

- total: 1 errors, 1 warnings, 9 lines checked
```

jshoen commented:
Sure.  What is right way to fix?  Do I need squash with the previous commit or can the fix have a separate commit and message on the branch?

andyross commented:
Squash please.  Zephyr main doesn't allow internal branching.

Commit:
kernel: kheap: fix k_heap_aligned_alloc handling of K_FOREVER

k_heap_aligned_alloc was not handling K_FOREVER timeout
correctly due to unsigned return value. Added explicit
K_FOREVER handling of end time.

Fixes #50611.

Signed-off-by: Jay Shoen <jay.shoen@perceive.io>

jshoen commented:
Ok.  Squashed and force pushed.  Hopefully all OK now.

fabiobaltieri commented on kernel/kheap.c:
```diff
@@ -66,6 +66,9 @@ void *k_heap_aligned_alloc(struct k_heap *h, size_t align, size_t bytes,
 	int64_t now, end = sys_clock_timeout_end_calc(timeout);
 	void *ret = NULL;
+
+	end = K_TIMEOUT_EQ(timeout, K_FOREVER) ? INT64_MAX : end;
```
Do we really like the ternary here? Not a fan, "else" case being a no-op makes one wonder if that was really the intent, this could just be

```c
if (K_TIMEOUT_EQ(timeout, K_FOREVER) {
        end = INT64_MAX;
}
```

Anyway, up to you folks.

andyross replied:
Personally I'm a fan of vertical compression; source code gets too long to read in most modern styles.  Semantically my brain reads that like MIN/MAX/CLAMP, where the if() signals "behavior".  But these are going to generate the same code, so I don't see much reason to care.

jshoen replied:
That being the case I prefer to leave as-is.

fabiobaltieri replied:
Cool with that, thanks for the quick turnaround folks!

andyross reviewed (approved):
refresh +1
